:sparkles: Add sphinx directives to generate envvar documentation

stevenbal opened the pull request:
Closes https://github.com/maykinmedia/open-api-framework/issues/83 partially

Example in docs: https://open-klant--498.org.readthedocs.build/en/498/installation/config/env_configuration.html

stevenbal commented on pyproject.toml:
```diff
@@ -79,6 +79,14 @@ otel = [
     "opentelemetry-instrumentation-django",
     "opentelemetry-resource-detector-containerid",
     "opentelemetry-sdk",
+    "python-decouple>=3.8",
```
I guess if we will use `config` in other modules, this should be part of the base dependencies?

sergei-maertens replied:
Dotenv files are only used in development, in prod they are injected as real environment variables.

Python dotenv loads the .env file and injects the parsed envvars in is environ. Python decouple doesn't, it treats the .env file as if it's is environ, but the OTEL SDK only reads from os.environ.

So any envvars you put in .env with only python-decouple will not be seen by tooling that only reads from is environ, which is *all* of our tooling.

sergei-maertens commented on maykin_common/otel.py:
```diff
@@ -54,6 +55,38 @@
+#         "Don't do this on Kubernetes, as it may lead to conflicting information."
+#     ),
+# )
+# PROTOCOL: ExportProtocol = config(
+#     "OTEL_EXPORTER_OTLP_PROTOCOL",
+#     default=DEFAULT_PROTOCOL,
+#     group="Open Telemetry",
+#     help_text=(
+#         "The name of the protocol to use for OTel, possible options: ``grpc``, "
+#         "``http/protobuf``."
+#     ),
+# )
```
I would also not actively document the OTel environment variables, because they are standardized across langauges and SDK by the OTel standard itself. Linking to https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#general-sdk-configuration should be sufficient. Maintaining a copy of this documentation can only go stale and be wrong eventually, and it does not produce any worthwile benefits IMO.

stevenbal committed:
:sparkles: Add sphinx directives to generate envvar documentation

stevenbal committed:
:memo: Do not document OTel envvars explicitly via generated docs

stevenbal committed:
:white_check_mark: Add tests for envvar documentation sphinx directives

stevenbal committed:
:memo: Add docs for envvar documentation sphinx directives

stevenbal committed:
:green_heart: Ignore tox dir when compiling messages before tests

because the translation files for sphinx seem to contain errors, causing the command and job to crash:

Example error: .tox/py312-django52-base/lib/python3.12/site-packages/sphinx/locale/fa/LC_MESSAGES/sphinx.po:599: a format specification for argument 'default.__name__' doesn't exist in 'msgstr'
msgfmt: found 1 fatal error

stevenbal commented on docs/env_docs_helpers.rst:
```diff
@@ -0,0 +1,90 @@
+        help_text=(
+            "a comma separated (without spaces!) list of domains that serve "
+            "the installation. Used to protect against Host header attacks."
+        ),
```
As discussed earlier, the setup with docstrings is problematic because setting variable names and envvar names must exactly match, so I left the setup as is

stevenbal commented on maykin_common/documentation/config_directives.py:
```diff
@@ -0,0 +1,247 @@
+        return [v for v in variables if v.name in members]
+    if exclude:
+        return [v for v in variables if v.name not in exclude]
+    return variables
```
I changed it to use iterators instead, though I'm not sure about combining `members` and `exclude`, because it seems a bit weird to specify both? Also I would then have to instantiate initial list of variables twice, because if `members` = [], doing `members - exclude` would result in an empty set

sergei-maertens replied:
Agree that it seems weird, though it shouldn't be impossible as long as you have a well-defined precedence order. Okay to keep it as is.

sergei-maertens commented on maykin_common/config.py:
```diff
@@ -86,6 +142,10 @@ def config[T](
+    help_text: str = "",
+    group: str | None = None,
+    add_to_docs: bool = True,
+    auto_display_default: bool = True,
```
I would heavily lean towards putting that information in a single `documentation` parameter which is its own data class. As you see int his PR, the combinatorial explosion on the typing overloads is unwieldly. That allows you to modify and extend that API in the future without affecting everything else in the process too.

The documentation configuration is a completely different thing from the actual runtime configuration and that should be made clear in the API IMO.

stevenbal committed:
:recycle: Add single documentation parameter for config helper

to make sure the documentation parameters are contained within a single dataclass

stevenbal commented on maykin_common/vcr.py:
```diff
@@ -160,7 +160,9 @@ def raise_exception(request):
-        return clean_vcr.use_cassette(self._get_cassette_name())
+        context_decorator = clean_vcr.use_cassette(self._get_cassette_name())
+        assert isinstance(context_decorator, CassetteContextDecorator)
+        return context_decorator
```
This was failing suddenly, seems to be unrelated to the changes: https://github.com/maykinmedia/django-common/actions/runs/24393254422/job/71244511451#step:6:28

sergei-maertens commented on maykin_common/config.py:
```diff
@@ -1,23 +1,52 @@
+        documentation.
+    auto_display_default : bool, default ``True``
+        Indicates whether the documentation directives should display the specified
+        default. Can be set to ``False`` if you want to manually specify a default.
```
Can these be added as docstrings below each parameter instead? I'm not sure if Sphinx will pick them up correctly. I'm worried about the duplication of type information now that can diverge over time.

sergei-maertens commented on maykin_common/config.py:
```diff
@@ -1,23 +1,52 @@
+
+
+"""Shorthand to not include an environment variable in generated documentation"""
+no_doc = DocumentationParams(add_to_docs=False)
```
```suggestion
no_doc = DocumentationParams(add_to_docs=False)
"""Shorthand to not include an environment variable in generated documentation"""
```

sergei-maertens commented on maykin_common/config.py:
```diff
@@ -117,8 +168,27 @@ def config[T](
+    The ``documentation`` parameter is available to generate documentation for
+    environment variables with Sphinx directives provided by this library:
+
+    documentation (:class:`maykin_common.config.DocumentationParams`)
```
```suggestion
    :param documentation: See :class:`maykin_common.config.DocumentationParams`
```
for proper Sphinx markup/documentation?

sergei-maertens commented on maykin_common/config.py:
```diff
@@ -117,8 +168,27 @@ def config[T](
+    documentation (:class:`maykin_common.config.DocumentationParams`)
     """
 
+    if not documentation:
+        # Instantiate the defaults
+        documentation = DocumentationParams()
+
+    if documentation.add_to_docs:
```
```suggestion
    if documentation and documentation.add_to_docs:
```
or am I missing something?

sergei-maertens replied:
ah, possibly that if nothing is specified, it will be included by default?

stevenbal replied:
yeah that was my thought indeed, with the current setup it includes everything by default (at least in the registry). Do you disagree with that approach?

sergei-maertens replied:
nope, let's go for it :) (PR is approved ;) )

sergei-maertens reviewed (approved):
Some suggestions that you can overrule if you think I'm nitpicking too much 😄

stevenbal committed:
:memo: Slight fixes for config docs helpers docstrings
